docs/LibraryEvolutionManifesto.md

```diff
@@ -0,0 +1,215 @@
+
+```swift
+// Library, version 1
+protocol MagicType {}
+protocol Wearable {}
+func use<T: MagicType>(_ item: T) {}
+```
+
+```swift
+// Client, version 1
+struct Amulet : MagicType, Wearable {}
+use(Amulet())
+```
+
+```swift
+// Library, version 2
+protocol MagicType {
+  @available(dishwasherOS 2.0, *)
+  func equip()
+}
+extension MagicType {
+  @available(dishwasherOS 2.0, *)
+  func equip() { print("Equipped.") }
+}
+
+protocol Wearable {}
+extension Wearable where Self: MagicType {
+  @available(dishwasherOS 2.0, *)
+  func equip() { print("You put it on.") }
+}
+
+func use<T: MagicType>(_ item: T) { item.equip() }
+```
+
+Let's say we're running dishwasherOS 2.0. Before the client is recompiled, the implementation of `equip()` used for `Amulet` instances can only be the default implementation, i.e. the one that prints "Equipped". However, recompiling the client will result in the constrained implementation being considered a "better" match for the protocol requirement, thus changing the behavior of the program.
+
+This should never change the *meaning* of a program, since the default implementation for a newly-added requirement should always be *correct.* However, it may have significantly different performance characteristics or side effects that would make the difference in behavior a surprise.
+
+This is similar to adding a new overload to an existing set of functions, which can also change the meaning of client code just by recompiling. However, the difference here is that the before-recompilation behavior was never requested or acknowledged by the client; it's just the best the library can do.
+
+A possible solution here is to require the client to acknowledge the added requirement in some way when it is recompiled.
+
+(We do not want to perform overload resolution at run time to find the best possible default implementation for a given type.)
+
+
+# Evolving Attributes
+
+A number of annotations (attributes and modifiers) would benefit from being able to change between releases of a library. For example:
+
+- Making a `public` class `open`.
+- Making a `public` class `final`.
+- Making a struct or enum `@frozen`.
+- Making a function `@inlinable`.
+
+However, all of these changes have to be associated with a specific release to be correct:
+
+- The newly-`open` class can't be subclassed if the client's deployment target is too old.
+- Convenience initializers on a newly-`final` class can't satisfy protocol requirements if the client's deployment target is too old.
+- A newly-`@frozen` struct may have had more private stored properties in the past.
+- A newly-`@inlinable` function's body may not work correctly with older versions of the library.
+
+While solutions can be tailored to each of these problems, they have a common need to record the OS version when an annotation is added or changed. Without that, we're locked in to the way a declaration is originally published, or having to take correctness on faith.
+
+## Limitations
```

Comment:
@Gankra Here's the "can't 100% retroactively `@frozen`" thing I was talking about yesterday.